File: app/src/main/java/com/example/jonathan/webjsonviewer/adapter/CustomerAdapter.java
package com.example.jonathan.webjsonviewer.adapter;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.jonathan.webjsonviewer.R;
import com.example.jonathan.webjsonviewer.model.Customer;

import java.util.ArrayList;

public class CustomerAdapter extends RecyclerView.Adapter<CustomerAdapter.CustomerViewHolder> {
  private static final String TAG = "WJV " + CustomerAdapter.class.getSimpleName();

  private final ArrayList<Customer> mCustomerList;

  public CustomerAdapter(ArrayList<Customer> customerList) {
    Log.d(TAG, "CustomerAdapter");

    mCustomerList = customerList;
  }

  @NonNull
  @org.jetbrains.annotations.NotNull
  @Override
  public CustomerViewHolder onCreateViewHolder(@NonNull @org.jetbrains.annotations.NotNull ViewGroup parent, int viewType) {
    Log.d(TAG, "onCreateViewHolder");

    LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
    View view = layoutInflater.inflate(R.layout.row_customer, parent, false);
    return new CustomerViewHolder(view);
  }

  @Override
  public void onBindViewHolder(CustomerViewHolder holder, int position) {
    Log.d(TAG, "onBindViewHolder");

    holder.txtCusName.setText(mCustomerList.get(position).getName());
    holder.txtCusEmail.setText(mCustomerList.get(position).getEmail());
    holder.txtCusBalance.setText(mCustomerList.get(position).getBalance());
    holder.txtCusFriends.setText(String.valueOf(mCustomerList.get(position).getFriends().size()));
  }

  @Override
  public int getItemCount() {
    Log.d(TAG, "getItemCount: " + mCustomerList.size());

    return mCustomerList.size();
  }

  static class CustomerViewHolder extends RecyclerView.ViewHolder {
    String TAG = "WJV " + CustomerViewHolder.class.getSimpleName();

    final TextView txtCusName;
    final TextView txtCusEmail;
    final TextView txtCusBalance;
    final TextView txtCusFriends;

    CustomerViewHolder(View itemView) {
      super(itemView);

      Log.d(TAG, "CustomerViewHolder");

      txtCusName = (TextView) itemView.findViewById(R.id.txt_customer_name);
      txtCusEmail = (TextView) itemView.findViewById(R.id.txt_customer_email);
      txtCusBalance = (TextView) itemView.findViewById(R.id.txt_customer_balance);
      txtCusFriends = (TextView) itemView.findViewById(R.id.txt_customer_friends);
    }
  }
}

File: app/src/main/java/com/example/jonathan/retrofitwebjsonviewer/adapter/CustomerAdapter.kt
package com.example.jonathan.retrofitwebjsonviewer.adapter

import android.util.Log
import com.example.jonathan.retrofitwebjsonviewer.model.Customer
import androidx.recyclerview.widget.RecyclerView
import com.example.jonathan.retrofitwebjsonviewer.adapter.CustomerAdapter.CustomerViewHolder
import com.example.jonathan.retrofitwebjsonviewer.adapter.CustomerAdapter
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import com.example.jonathan.retrofitwebjsonviewer.R
import android.widget.TextView
import java.util.ArrayList

// After auto conversion from Java to Kotlin, "internal" has to be added:
internal class CustomerAdapter(customerList: ArrayList<Customer>) :
    RecyclerView.Adapter<CustomerViewHolder>() {
    private val mCustomerList: ArrayList<Customer>

    init {
        Log.d(TAG, "CustomerAdapter")
        mCustomerList = customerList
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): CustomerViewHolder {
        Log.d(TAG, "onCreateViewHolder")
        val layoutInflater = LayoutInflater.from(parent.context)
        val view = layoutInflater.inflate(R.layout.row_customer, parent, false)
        return CustomerViewHolder(view)
    }

    override fun onBindViewHolder(holder: CustomerViewHolder, position: Int) {
        Log.d(TAG, "onBindViewHolder")
        holder.txtCusName.text = mCustomerList[position].name
        holder.txtCusEmail.text = mCustomerList[position].email
        holder.txtCusBalance.text = mCustomerList[position].balance
        holder.txtCusFriends.text = mCustomerList[position].friends.size.toString()
    }

    override fun getItemCount(): Int {
        Log.d(TAG, "getItemCount: " + mCustomerList.size)
        return mCustomerList.size
    }

    internal class CustomerViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var TAG = "WJV " + CustomerViewHolder::class.java.simpleName
        val txtCusName: TextView
        val txtCusEmail: TextView
        val txtCusBalance: TextView
        val txtCusFriends: TextView

        init {
            Log.d(TAG, "CustomerViewHolder")
            txtCusName = itemView.findViewById<View>(R.id.txt_customer_name) as TextView
            txtCusEmail = itemView.findViewById<View>(R.id.txt_customer_email) as TextView
            txtCusBalance = itemView.findViewById<View>(R.id.txt_customer_balance) as TextView
            txtCusFriends = itemView.findViewById<View>(R.id.txt_customer_friends) as TextView
        }
    }

    companion object {
        private val TAG = "WJV " + CustomerAdapter::class.java.simpleName
    }
}